### PR TITLE
Add validations to the endpoints

### PR DESCRIPTION
**Pull request by duncanjbrown**
### Context

When a Vendor posts invalid data to our endpoints (as they will), we need to return suitable and useful errors.

### Changes proposed in this pull request

- [X] 0. Remove the `rejected_at` field from the `ApplicationChoice` model. This field is not in line with our plan of using the audit log to assign dates to events. At least, it's not clear that this is the implementation we want. As keeping this would involve adding more code to the service at this stage, I've taken it out.
- [X] 1. Add validations to the `MakeAnOffer` service and surface them via the API: max 20 conditions, and max 255 chars per condition. allow nil for unconditional offers.
- [X] 2. Add validations to the `Reject` service and surface them via the API: require rejection_reason and limit it to 255 chars

I was able to use built-in validations for rejection_reason but not for offer_conditions bc it's an array.

See commit messages for more details.

### Guidance to review

There are no validations in the `ApplicationChoice` model. How do we feel about that? Note that JSON fields require somewhat special treatment.

### Link to Trello card

https://trello.com/c/2wtVwvIj/1077-add-validations-to-the-endpoints

### CODE REVIEW

File: app/services/reject_application.rb [hunk 1]
```diff
@@ -1,18 +1,24 @@
 class RejectApplication
+  attr_reader :rejection_reason
+
   include ActiveModel::Validations
 
-  def initialize(application_choice:, rejection:)
+  validates_presence_of :rejection_reason
+  validates_length_of :rejection_reason, maximum: 255
```
Comment (tvararu) [inline app/services/reject_application.rb]: Interesting, I'm using `validates:` in my parallel PR for the Personal details validations, e.g.:

```ruby
    validates :first_name, :last_name, :english_main_language,
              :first_nationality,
              presence: true

    validates :first_name, :last_name,
              length: { maximum: 100 }
```

I'm not sure if my branch passes Rubocop yet (have some trouble running the bugger locally), but do we want consistency on this? Tbh I'm fine with mixing the two kinds of sugar in our codebase, but if we want 1 flavour I'm happy to switch to this vice/versa etc

File: spec/requests/vendor_api/post_make_an_offer_spec.rb [hunk 2]
```diff
@@ -53,6 +53,20 @@
     expect(parsed_response).to be_valid_against_openapi_schema('UnprocessableEntityResponse')
   end
 
+  it 'returns an error when given invalid conditions' do
+    application_choice = create(:application_choice, status: 'application_complete')
+
+    post_api_request "/api/v1/applications/#{application_choice.id}/offer", params: {
+      data: {
+        conditions: 'DO NOT WANT',
```
Comment (tvararu) [inline spec/requests/vendor_api/post_make_an_offer_spec.rb]: Do not want

File: app/services/reject_application.rb (hunk 1 above)
Comment (duncanjbrown) [reply app/services/reject_application.rb]: No strong feelings — I think I prefer the `validates` version as it naturally groups rules by field. happy to conform

Comment (tvararu) [reply app/services/reject_application.rb]: No strong feelings either so feel free to leave as is, it's all readable to me 👍

File: app/services/make_an_offer.rb [hunk 3]
```diff
@@ -20,4 +24,18 @@ def save
     errors.add(:state, e.message)
     false
   end
+
+private
+
+  def offer_conditions_valid?
```
Comment (alexdesi) [inline app/services/make_an_offer.rb]: The name of this methods implies that it check whether `makeanoffer` is valid, but it isn't the case.
`valid?`(line 12) is the one that does it, and it just checks if `errors` is empty? 

Should this be renamed to `validate_condition`? (or similar)

File: spec/services/make_an_offer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+require 'rails_helper'
+
+RSpec.describe MakeAnOffer do
+  describe 'validation' do
+    it 'accepts nil conditions' do
+      decision = MakeAnOffer.new(
+        application_choice: create(:application_choice, status: :application_complete),
```
Comment (alexdesi) [inline spec/services/make_an_offer_spec.rb]: this way we are creating 4 times the
`create(:application_choice, status: :application_complete`.

I'm in favour of using a let in these cases:

`let(:completed_application_choice) { create(:application_choice, status: :application_complete }`

apart from performance, it is good for readability as well.

File: app/services/make_an_offer.rb [hunk 5]
```diff
@@ -20,4 +24,18 @@ def save
...
+    return if @offer_conditions.blank?
+
+    unless @offer_conditions.is_a?(Array)
+      errors.add(:offer_conditions, 'must be an array')
+      return
+    end
+
+    errors.add(:offer_conditions, 'has over 20 elements') if @offer_conditions.count > 20
+    errors.add(:offer_conditions, 'has a condition over 255 chars in length') if @offer_conditions.any? { |c| c.length > 255 }
+  end
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/services/make_an_offer.rb]: :+1: Good idea

File: app/services/make_an_offer.rb (hunk 3 above)
Comment (duncanjbrown) [reply app/services/make_an_offer.rb]: I don't think I follow — wouldn't `make_an_offer_valid?` imply whether this `MakeAnOffer` is valid? If this is confusing, though, it should be changed

File: spec/services/make_an_offer_spec.rb (hunk 4 above)
Comment (duncanjbrown) [reply spec/services/make_an_offer_spec.rb]: No strong feelings about the style — please feel free to fix this up if you like. `let`s are not cached across examples, so I don't think there's a performance benefit to be had here.

**Comment by duncanjbrown** [comment]
Going to add some `CONSTANTS`

**Comment by duncanjbrown** [comment]
The constants have been added

File: spec/services/make_an_offer_spec.rb (hunk 4 above)
Comment (alexdesi) [reply spec/services/make_an_offer_spec.rb]: @duncanjbrown thanks for reminding me the about let caching :)
I think @tijmenb suggestion is a good approach.
ps.
I think there are others way to cache the application_choice, (`before :all` for example) but probably at the expense of readability.

File: app/services/make_an_offer.rb (hunk 3 above)
Comment (alexdesi) [reply app/services/make_an_offer.rb]: `validate_offer_conditions` sounds good!

File: spec/services/make_an_offer_spec.rb (hunk 4 above)
Comment (duncanjbrown) [reply spec/services/make_an_offer_spec.rb]: Thanks @tijmenb: fixed up with `build_stubbed`

**Pull request merged**